### ISSUE TITLE
Add support for RML import format

### ISSUE THREAD

**Issue by pietervaneverdingen** [pldn/LDWizard#60]
### User story "Pick up where somebody left of"

- [Who] As a user of LD Wizard,
- [What] I want to be able to import an RML file that was exported in a previous LD Wizard session,
- [Why] So that I can improve a transformation that was created in the past (instead of having to start over).

### Background

- LD Wizard can be used to export an RML transformation to file.
- It is currently not possible to reuse this RML file later to change the transformation configuration inside LD Wizard.
- As a result, changing an existing transformation can only be achieved by recreating it from scratch.

### Acceptance criteria

- Support loading an RML file next to a CSV file.
- The RML file is assumed to be exported during a previous LD Wizard session.  (RML files that have a different origin are not supported.)
- When moving to the second page in LD Wizard ("Configure"), the GUI should be configured in accordance with the transformations that are specified in the loaded RML file (e.g. property mappings).

---

A [bounty](https://www.pldn.nl/wiki/Bounties_List) has been placed on this issue by [PLDN](https://www.pldn.nl) for € 2000
 
[Click here to learn more](https://www.pldn.nl/wiki/Bounty_Procedures) if you're interested in claiming this bounty by resolving this issue.

**Comment by pietervaneverdingen** [comment]
The scope of this issue will be enlarged to also include "Add support for YARRRML import format".

And the bounty sum for this issue will be increased with € 500 to € 2500

**Comment by vemonet** [comment]
Hi @pietervaneverdingen I would be interested to look into implementing this feature, but there is a blocking issue with how LDWizard handles "column refinements"

## Problem

Column refinements are small functions defined by a LDWizard owner that enable the users to add additional transformation to the content of a cell (e.g. add a prefix to create a URI from a plain string). But they have been implemented to be applied before the RML transformation. Which means the infos about "columns refinements" are not present in the RML mappings

So for a transformation with columns refinements if you provide your old CSV + the RML that was previously generated, we will not be able to figure out what "column refinements" have been applied. Because the information is just not there

## Solutions?

1. The ideal solution would be to migrate from the old "column refinements" system to use custom RML functions, RocketRML makes it easy to define them, they could be defined by each LD wizard owner in the `ldwizard_config.ts` file similarly to how we do with column refinements at the moment 

Then instead of running the column refinement ourselves we would just add the custom function to the RML, and RocketRML will take care of running the custom function for us (less code to maintain in LDWizard 🍾)

Imo that is the most logical and standard-compliant solution. But it is a **major change** to the whole LDWizard mechanism, and it might be a bit of a pandora box... It should be relatively easy (99% of the job already done by RocketRML), but not sure how much work it will need to get to a better state than the existing solution (will the RocketRML custom functions will cover all use-cases covered by the Column refinement?)

2. The quick solution would be to implement the import RML feature and ignore columns refinements. But then the feature will be incomplete (and completely broken for all use-case with column refinements).


## Conclusion

In my opinion if you want to be serious about using RML as shareable mapping format, then it needs to be fully supported, not partially. So the LDWizard should work directly with RML custom functions instead of having their own column refinement system in parallel

What are your thoughts on this? Are there specific needs that make the current column refinement system required, and prevent us to use RML custom functions?

**Comment by vemonet** [comment]
I looked a bit more into it and have some amends to do to my previous message!

1. I forgot that the LDWizard provide the CSV after applying Column Refinments at the last step, so if the user reuse the RML mappings + the CSV after refinements, then the mappings can be imported without issue
2. RML mappings with custom functions will only work on a RML engine that have implemented those custom functions. So only the RocketRML engine of the LDWizard used to do the mappings will be able to run the RDF generation (apart if you reimplement those custom functions for other RML engines manually)

Considering those 2 points, continuing to use the current columns refinements makes more sense as it makes the system more reusable, even if it is not 100% following the RML specifications

I will look into adding import and YARRRML support when I have some time then :)

**Comment by pietervaneverdingen** [comment]
Hi @vemonet, many thanks for your proposal and additional refinement. 

It is best to discuss your ideas with LDWizard Gatekeeper @mightymax first to decide on the best-possible scenario. 

And then it is a good idea to schedule an online LDWizard session with the RML experts in our LDWizard working group on the short term, to finalize the preferred scenario that we all have the same understanding of what can be developed within what timeframe and how it can fit with the other LDWizard development activities.

Then we can assign the bounty to you and finish the development activities with all the relevant stakeholders involved.

**Comment by vemonet** [comment]
Thanks @pietervaneverdingen, the instructions and features to implement were clear, so I went ahead and implemented them already:
* Export YARRRML 
* Import RML (includes support for IRI Prefix transformation column refinements)
* Import YARRRML 

The UI impact is minimal, I just added a "mapping file upload" button under the CSV upload button at the first step (which is optional, users can still just upload a CSV alone)

Importing is working well for the different mappings I generated

There is just one small issue: when the option "Row number" is chosen for the Key Column, then generating YARRRML fails. Because we use blank nodes for the SubjectMap, but YARRRML expect we use an URI.
This problem has been mentioned in this issue: https://github.com/pldn/LDWizard/issues/144 (where I gave more details about the problem)
I think there will need to be a small discussion between the LDWizard stakeholders to decide which way to take for row numbers (do we continue to use blank nodes? or do we do the effort of adding row number support to RocketRML to be able to have URI as SubjectMap?)

I missed last week session, but I can join the next one to present the implementation! Discuss if it works as expected for you, and what you would like me to change

**Comment by vemonet** [comment]
A bit more information about the implementation, for those who might try it:

* The "Import mappings" button shows up once the user has loaded a CSV file. User can load RML or YARRRML mappings, the LDWizard will automatically detect which parser to use
* YARRRML files are converted to RML using the "official" [`yarrrml-parser`](https://github.com/RMLio/yarrrml-parser/) npm package, and we do the import from the RML mappings
* It supports importing RML mappings from turtle format only at the moment. As far as I know RML mappings are shared using turtle most of the time, and it is the format we use to export the mappings in the LDWizard. I could easily add support for ntriples, trig, quads through the N3.js package that is already installed, but we would need to install other packages if we want to also support for more exotic formats like JSON-LD or XML.
* Standard IRI transformation are imported, but, as mentionned previously, custom column refinements cannot be imported from RML, so we expect the users will use the transformed CSV that was exported during the last LDWizard mapping session
*  Since YARRRML doesn't support blank nodes for SubjectMap, when the option "Row number" is chosen for the Key Column, the YARRRML export option is not shown to the user (they can only export to RML and CSVw as before)

**Comment by philipperenzen** [comment]
Hi @vemonet 

We recently did an update to a version 3 and have added some new features to LDWizard.

Quite a lot of changes have been made in the RML script, and in the navigation buttons during the SHACL integration #59. I believe these aforementioned changes might have some impact on your forked repository, most of the other features should not however. Please let me know if you run into any problems when rebasing.

**Comment by vemonet** [comment]
Thanks for the notification @philipperenzen, I rebased in a new branch: https://github.com/vemonet/LDWizard/tree/add-import-and-yarrrml-rebased

That also reminded me I took the opportunity to also fix some small issues alongside adding import:
* Improved the RDF generation to output turtle instead of ntriples (https://github.com/pldn/LDWizard/issues/34)
* Fixed example in `CONFIGURING.md` (https://github.com/pldn/LDWizard/issues/148)
* Added a few missing config parameters to the `runtimeConfig.ts` so that it presents most parameters available to the user (type annotation and autocomplete is nice, but for this kind of config the best from a dev point of view is to have a complete example to start from)

I can present it tomorrow during the LDWizard call

**Comment by pietervaneverdingen** [comment]
Hi @vemonet, thanks for the update! Daniel is also available tomorrow to demo his SHACL-support functionality. I suggest that Daniel demos the functionality that he has built first and after that you can then present and demo your work.

**Comment by vemonet** [comment]
Perfect for me, I was hoping to discover the new SHACL functionality!

**Comment by mightymax** [comment]
@pietervaneverdingen is this issue still open and eligible  for [the €2.500 bounty](https://www.pldn.nl/wiki/Bounties_List)? All participants at the HackaLoD received an email pointing to the bounties, but if I remembered correctly @vemonet fixed this and the issue is now closed.

**Comment by vemonet** [comment]
@mightymax I sent an invoice for the €2.500 some months ago to info@pldn.nl and info@bdvc.nl , but never got an answer

I am sending it again right now

**Comment by pietervaneverdingen** [comment]
@vemonet apologies, I was not aware of this, I will contact the person at PLDN that looks after our financial administration to check with him where it went wrong.

**Comment by mightymax** [comment]
@pietervaneverdingen can you change the [Bounty list page](https://www.pldn.nl/wiki/Bounties_List) as well? I do not have the rights to do this.

**Comment by pietervaneverdingen** [comment]
@mightymax, done.
